Implement tcp fast retransmit

ssharks opened the pull request:
To improve the TCP performance in case of occasional packet loss implement the fast re-transmit algorithm. This solves issue https://github.com/zephyrproject-rtos/zephyr/issues/46351 and https://github.com/zephyrproject-rtos/zephyr/issues/47240

ssharks committed:
net: tcp: Send a duplicate ACK for out of order data

When out of order data is received, send out a duplicate ACK to notify
the sender that there is data received out of sequence, so it can
retransmit the missing section.

Also avoid sending acknowlegdement to acknowlegdement messages that are
having an out of order sequence number.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Implement a fast retransmit algorithm

Instead of waiting for the retransmit timeout, retransmit as soon as
missing data is deduced based on a triple-duplicate ACK.

Increase the number of buffers in the testcase, to allow for at least 4
packets in flight to trigger the triple-duplicate ACK.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Update testcase for acknowledging out of order packets

The stack acknowledges out of order packets right now, update the
test_server_recv_out_of_order_data test to properly validate this
behavior.

Out of order packets should be acknowledged using sequence number of
the last correctly received byte.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
Let's put it this way: It's a good thing the payload is not checked in static void test_server_recv_out_of_order_data(void), using the seq as a pointer probably not returning the data the tester intended.
,

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -2157,12 +2157,18 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				verdict = tcp_data_received(conn, pkt, &len);
 			} else if (net_tcp_seq_greater(conn->ack, th_seq(th))) {
-				tcp_out(conn, ACK); /* peer has resent */
+				if (len > 0) {
```
Do we need to check the SYN flag here?

ssharks replied:
Fixed

ssharks commented:
Does it make sense increasing the default CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT to a value larger then 100 ms? What value would make sense?

ssharks committed:
net: tcp: Restructure out of order testcase

To provide more fine grained testing on out of order acknowledgements
restructure the testcase to use a list with expected in and outputs.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
While making the test I bumped into the following issue: https://github.com/zephyrproject-rtos/zephyr/issues/47240

ssharks committed:
net: tcp: Correctly handle partial duplicate out of order data

TCP is a streaming protocol, this means it can set the packet boundary
at an arbitrary location. Therefor a re-transmitted packet may have the
packet boundary at a different location from the original missing segment.

The reordering logic should be able to deal with this situation and
throw away the overlapping data.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Correctly handle partial duplicate buffered out of order data

TCP is a streaming protocol, this means it can set the packet boundary
at an arbitrary location. Therefor a re-transmitted packet may have the
packet boundary at a different location from the original missing segment.

The reordering logic should be able to deal with this situation and
throw away the overlapping data. Now also for data that needs to be added
to the existing queue

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
Actually open TODOs:
- Validate the data in the testcase upon reception
- Handle the situation where the incoming data is fully including the packet that is already queued, in that case clear the existing buffer and start over.

rlubos reviewed (commented):
Do I understand it correctly that this PR also fixes https://github.com/zephyrproject-rtos/zephyr/issues/47240? If so, please add `Fixes #47240` to the PR description so that the issue gets linked with the PR.

ssharks committed:
net: tcp: Correctly deal with fully duplicated buffered out of order data

TCP is a streaming protocol, this means it can set the packet boundary
at an arbitrary location. Therefor a re-transmitted packet may have the
packet boundary at a different location from the original missing segment.

The reordering logic should be able to deal with this situation and
throw away the overlapping data. This fix clears the out of order queue
when the queued data is not relevant.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -2157,12 +2157,18 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				verdict = tcp_data_received(conn, pkt, &len);
 			} else if (net_tcp_seq_greater(conn->ack, th_seq(th))) {
-				tcp_out(conn, ACK); /* peer has resent */
+				if ((len > 0) || FL(&fl, &, SYN)) {
```
Hmmm... I was not aware of this usecase, let's re-phase: how to differentitate between an ACK and a zero length packet, like a keep alive?
I looked at https://www.wireshark.org/docs/wsug_html_chunked/ChAdvTCPAnalysis.html#:~:text=TCP%20Keep%2DAlive%20ACK&text=The%20current%20sequence%20number%20is%20the%20same%20as%20the%20next,SYN%2C%20FIN%2C%20or%20RST, but did not get a very clear awnwer.

rlubos replied:
Hmm, I might not be seeing something, but is there a scenario when a proper empty ACK is sent with an older `SEQ` number? My understanding is that an older `SEQ` would be sent in only 2 cases:
* ACK from our side was lost, so the peer retransmits,
* or peer sends this keep-alive message, with an older `SEQ` number.

In either case, such a packet should be acknowledged. https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/overview.html exaplains that the keep-alive message specifically uses an older `SEQ` to assure ACK reply in any case, even if peer doesn't support keep-alive mechanism.

But perhaps I miss some corner case, when out-of-order data processing kicks in?

ssharks replied:
It is a though one. The risk is that the both ends of a connection end up in a situation where they keep sending ACKs on each others ACKs en you end up in an indefinite loop.

The only real reason I can think of when you will see out of order ACKs is when you will have bidirectional data flow at the same time. Then in the sending part a packet is being lost, the other side will be transmitting out of order ACKs.
Acking all the old ACKs does not pose too big of a risk for that reason.

There are only 1 special case where you want to ACK. Is a keep alive packet. The other closely related one in the ZWP. With ZWP I see most specifications state that you have to transmit a "dummy" byte. Although the current Zephyr implementation does modify the sequence number, but sends no data. It looks more like a keep alive packet for that reason.
The ACK behavior in this change will therefore not work with the ZWP for Zephyr.

The keep alive packets can only be arriving when the send_data_total = 0. Otherwise the condition will resolve itself anyway.
For the time being I'll revert this part of the change. But maybe it is at least useful to add a comment and create an issue for the ZWP behavior.

rlubos replied:
Regarding the Keep-alive packet, I don't think that sending an empty packet is incorrect, actually [RFC 1122](https://datatracker.ietf.org/doc/html/rfc1122#page-102) states that sending dummy data byte is rather an optional behaviour:
```
An implementation SHOULD send a keep-alive segment with no
data; however, it MAY be configurable to send a keep-alive
segment containing one garbage octet, for compatibility with
erroneous TCP implementations.
```

ssharks replied:
Indeed keep-alive packets should have a zero length, ZWP however should have a dummy byte.

rlubos replied:
Ok, that's the theory, we can fix the Zephyr implementation of ZWP sending, although it complicates it a bit (for example what to do if there is no actual data to send). Nonetheless, in the actual field I've observed that some servers send Keep-alive message as a ZWP (that's the reason it's implemented like this in Zephyr as well) so we need to be prepared to handle this case as well.

ssharks commented:
@jukkar / @rlubos: I'm a little bit stuck getting the contents of the data. I made an attempt in https://github.com/zephyrproject-rtos/zephyr/commit/8ade6b57d5bbec7d4e0a4a3561ed9d119ddbcd54, but in the tcp_recv_cb, it looks like the pkt is always NULL, so I'm not able to check the contents properly. Am I on the right track or how would you implement this?

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -2083,7 +2086,41 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+					/* Increment the duplicate acc counter,
+					 * but maximize the value
+					 */
+					conn->dup_ack_cnt = MIN(conn->dup_ack_cnt+1, 4);
```
Fixed

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -2083,7 +2086,41 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+			}
+
+			/* Only do fast retransmit when not already in a resend state */
+			if ((conn->data_mode == TCP_DATA_MODE_SEND) && (conn->dup_ack_cnt == 3)) {
```
Fixed

rlubos commented:
> @jukkar / @rlubos: I'm a little bit stuck getting the contents of the data. I made an attempt in [8ade6b5](https://github.com/zephyrproject-rtos/zephyr/commit/8ade6b57d5bbec7d4e0a4a3561ed9d119ddbcd54), but in the tcp_recv_cb, it looks like the pkt is always NULL, so I'm not able to check the contents properly. Am I on the right track or how would you implement this?

It seems to me that your approach is correct - if the callback function gets a NULL pkt pointer all the time it could indicate some underlying issue...

rlubos commented:
@ssharks Is the PR ready for final review or are you still working on it?

ssharks commented:
Although the theory is nice, I ran the large socket test, with Nagle disabled and bumped into the following issue:
ASSERTION FAIL [buf->len >= len] @ WEST_TOPDIR/zephyr/subsys/net/buf.c:1155, aparently the way I handle buffers at the moment is not working in all cases.

ssharks committed:
net: tcp: Also run the packet loss test with NO_DELAY option

When the NO_DELAY option is enabled, the retransmitted packets
will likely be differently broken down from the packets transmitted
initially. This provides a nice stress test for the out or order
reception logic.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
It took considerable effort to make the fast re-transmit work reliable under all conditions, The modification of the pending_len needs to be done carefully in the function tcp_check_pending_data due to the way the result is used afterwards.
I attempted to cover a lot of corner cases by both manual analysis as well as the unit test.

I believe the current implementation is functional, although there is room for improvement on clarity. Suggestions from reviewers are very welcome.

This makes me believe we must be lucky the CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT is default at 100 ms, which pretty much disables the out of order processing. Because the original implementation was easily going wrong.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -748,21 +749,57 @@ static size_t tcp_check_pending_data(struct tcp *conn, struct net_pkt *pkt,
 		struct tcphdr *th = th_get(pkt);
 		uint32_t expected_seq = th_seq(th) + len;
 		uint32_t pending_seq;
+		uint32_t gap_size;
```
Why can't the `gap_size` simply be an `int32_t` variable? It seems a bit of an overhead to me to have a separate pointer variable just to do the casting.

ssharks replied:
Could be, does that work in all cases correctly with overflowing?

rlubos replied:
It is no different than using a pointer, if the MSB is set the value will be interpreted as a negative. You may need to add an explicit cast to avoid warning (but we know what we're doing here):
`gap_size = (int32_t)(pending_seq - th_seq(th) - len);`

ssharks replied:
Thanks for the input, I applied your suggestion, with the added explicit cast of len to uint32_t

rlubos reviewed (commented):
Looks good, just one question.

ssharks commented:
Will be fixed on Wednesday

ssharks commented:
@jukkar: Could you have a look at this PR?

ssharks commented:
@jukkar/@rlubos Should we increase CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT default value from 100 ms to for example 1500 ms to have it more effective in practice? It is likely in practice there will be a timeout of 100 ms somewhere which renders basically disables the reordering process, because it had already timeout.

ssharks commented:
@AndreyDodonov-EH This might be interesting for you as well

rlubos reviewed (dismissed):
Thanks, LGTM.

rlubos commented:
> @jukkar/@rlubos Should we increase CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT default value from 100 ms to for example 1500 ms to have it more effective in practice? It is likely in practice there will be a timeout of 100 ms somewhere which renders basically disables the reordering process, because it had already timeout.

I think we can consider increasing the default value, however, we should carefully observe the results, as this would also imply increased buffer consumption.

ssharks commented:
@Andr

> > @jukkar/@rlubos Should we increase CONFIG_NET_TCP_RECV_QUEUE_TIMEOUT default value from 100 ms to for example 1500 ms to have it more effective in practice? It is likely in practice there will be a timeout of 100 ms somewhere which renders basically disables the reordering process, because it had already timeout.
> 
> I think we can consider increasing the default value, however, we should carefully observe the results, as this would also imply increased buffer consumption.

That is indeed a risk, the amount of stored data will never exceed the receive window, however due to fragmenting the number of used fragments can be larger then expected. This becomes worse the packets become smaller or the CONFIG_NET_BUF_DATA_SIZE becomes bigger. There is currently not a kind of defragmentation implemented or trimming of the end of the queue when the stack runs out of receive buffers.

ssharks commented:
@jukkar Could you look at this PR?

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1789,8 +1797,14 @@ static void tcp_queue_recv_data(struct tcp *conn, struct net_pkt *pkt,
+
+			if (offset < len) {
+				if (offset) {
+					net_buf_pull_mem(pkt->buffer, offset);
```
ditto

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -752,13 +752,20 @@ static size_t tcp_check_pending_data(struct tcp *conn, struct net_pkt *pkt,
+		pending_len = net_pkt_get_len(conn->queue_recv_data);
+		if (offset < pending_len) {
+			if (offset) {
+				net_buf_pull_mem(conn->queue_recv_data->buffer, offset);
```
I bumped into this issue in the packet loss testcase of the socket with no delay and replaced them by net_pkt_remove_tail. For this reason I added the no delay testcase. Are you commenting on the latest changes, or did I do something wrong?

jukkar reviewed (dismissed):
Hmm, I just wonder what happened here. I thought I was reading latest version of this PR but now I am not seeing those pull_mem calls any more. Anyway, PR looks ok now.

fabiobaltieri commented:
@ssharks can you rebase? There's a pending merge conflict.

carlescufi commented:
@ssharks please rebase

ssharks commented:
During rebase I bumped into and issue, need to investigate first,

ssharks commented:
Rebased the PR, but do not review jet

ssharks committed:
net: tcp: Discard duplicate data in reodering correctly

When all the data of the new packet is already present in the
queue_recv_data, make sure the packet is discarded.
This commit adds a test line that reproduces the issue and implements a
fix.

Also in appending the packet, call the official net_buf_frag_add function
instead of updating the last net_buf pointer, since the net_pkt_remove_tail
called in between, might have removed the last net_buf.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Add build in auto check for reordering

The reorder process in rather complicated. Add a runtime check to always
check afterwards if the reordering queue is still consistent. When this
test fails, discard the contents of the complete reorder buffer.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
I have found and fixed 2 remaining bugs. Also I had to do a rebase, with some merge conflicts so please carefully re-review. Especially around the ZWP exponential backoff.

There is always more to which for, but for now, I think this PR is ready for the time being.

On the which list:
- Merge the tcp_check_pending_data and tcp_queue_recv_data into a single one
- Support packets that are partially overlapping with already acknowledged data
- Support multiple disjoint out of order sequences

ssharks commented:
Sorry, another comment fix. Needed to update a comment after copy paste.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1810,18 +1810,31 @@ static bool tcp_validate_seq(struct tcp *conn, struct tcphdr *hdr)
 		(net_tcp_seq_cmp(th_seq(hdr), conn->ack + conn->recv_win) < 0);
```
Typo in commit message of `net: tcp: Add build in auto check for reordering`
s/conistend/consistent

ssharks replied:
Fixed

jukkar replied:
Hmm, now the commit msg says `conistent` should be `consistent`

ssharks replied:
Sorry, fixed now

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1810,18 +1810,31 @@ static bool tcp_validate_seq(struct tcp *conn, struct tcphdr *hdr)
 
+		if (last != NULL) {
+			if (next_seq != seq) {
+				NET_ERR("Incorrect order in out of order sequence");
```
Perhaps print this one in the calling function so that we could print the connection pointer too (might help when debugging issues)

ssharks replied:
Good suggestion, applied

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -1932,13 +1944,15 @@ static void tcp_queue_recv_data(struct tcp *conn, struct net_pkt *pkt,
+			NET_DBG("All pending data: conn %p", conn);
+			if (check_seq_list(conn->queue_recv_data->buffer) == false) {
+				/* error in sequence list, drop it */
+				net_buf_unref(conn->queue_recv_data->buffer);
```
So here we could print something like this
```
NET_ERR("Incorrect order in out of order sequence, conn %p", conn);
```

ssharks replied:
Indead

jukkar reviewed (commented):
Looks reasonable but this part of functionality is difficult to review, best is to have unit tests (which you have :+1:) and running the beast . Couple of minor comments found.

rlubos reviewed (dismissed):
LGTM, thanks for the effort. Seems like there is now a conflict to resolve in the test cases due to the recent ZTEST framework update (need to comply with the new test declaration API now).

ssharks commented:
@rlubos: Thanks for your review, I fixed the merge conflict in commit: https://github.com/zephyrproject-rtos/zephyr/pull/47115/commits/54956a72b346e85565c18086b8f25ec49f8a1ee4. The conflict was minor.

ssharks commented:
@jukkar: Are you satisfied with the implemented changes?
